### PR TITLE
Adding CRUISE + removing MIPAV and JIST dependencies

### PR DESCRIPTION
**Pull request by piloubazin**
* build-nodependencies.sh is the new option for building cbstools alone. It requires a list of specific java classes to be compiled (it might work without it, but it is safer that way). This list needs to be updated as we add new modules.

* cruise_cortex_extraction.py is now complete (and works faster and possibly better than the original).The corresponding JIST module has been named OptimCRUISE, to separate it from the previous one (as the code is different)

* embedded_syn.py runs ANTS and produces coordinate mappings as outputs, rather than collections of transformations (more robust that way). 100% python-based, with Nipype as the interface for ANTS i/o.

### CODE REVIEW

File: nighres/registration/embedded_syn.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    medium_iterations: float
+        Number of iterations at the medium level (default is 50)
+    fine_iterations: float
+        Number of iterations at the fine level (default is 40)
+    run_affine_first: bool
+        Runs a step of affine registration before the non-linear step (default is False)
+    cost_function: {'CrossCorrelation', 'MutualInformation'}
+        Cost function for the registration (default is 'MutualInformation')
+    interpolation: {'NearestNeighbor', 'Linear'}
+        Cost function for the registration (default is 'NearestNeighbor')
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: Interpolation for the registration

File: nighres/registration/embedded_syn.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    Returns
+    ----------
+    dict
+        Dictionary collecting outputs under the following keys
+        (suffix of output files in brackets)
+
+        * deformed_source (niimg): Deformed source image (_esyn_def)
+        * mapping (niimg): Coordinate mapping from source to target (_esyn_map)
+        * inverse (niimg): Inverse coordinate mapping from target to source (_esyn_invmap) 
+
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: What do you think about "syn" instead of "esyn"?

File: nighres/registration/embedded_syn.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    Notes
+    ----------
+    Port of the CBSTools Java module by Pierre-Louis Bazin. The main algorithm
+    is part of the ANTs software by Brian Avants and colleagues [1]_
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: We should mention that this is based on the Nipype interface and cite Nipype too

File: nighres/registration/embedded_syn.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+#from nipype.interfaces.ants import WarpTimeSeriesImageMultiTransform
+from nipype.interfaces.ants import ApplyTransforms
+
+# cbstools and nighres functions
+import cbstools
+from ..io import load_volume, save_volume
+from ..utils import _output_dir_4saving, _fname_4saving, \
+                    _check_topology_lut_dir
+
+# convenience labels
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: Why not use 0,1,2,3 below? Less convoluted. Or put these inside the function

File: nighres/registration/embedded_syn.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        for y in xrange(nsy):
+            for z in xrange(nsz):
+                src_coord[x,y,z,X] = x
+                src_coord[x,y,z,Y] = y
+                src_coord[x,y,z,Z] = z
+    src_map = nb.Nifti1Image(src_coord, source.affine, source.header)
+    src_map_file = _fname_4saving(file_name=file_name,
+                            rootfile=source_image,
+                            suffix='tmp_srccoord')
+    save_volume(os.path.join(output_dir, src_map_file), src_map)
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: output_dir is an optional input, what if it is not given? maybe save to the tmp folder instead?

File: nighres/registration/embedded_syn.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    trg_map_file = _fname_4saving(file_name=file_name,
+                            rootfile=source_image,
+                            suffix='tmp_trgcoord')
+    save_volume(os.path.join(output_dir, trg_map_file), trg_map)
+    
+    # run the main ANTS software
+    ants = ANTS()
+    ants.inputs.dimension = 3
+    
+     # add a prefix to avoid multiple names?
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: can you explain?

File: nighres/utils.py [hunk 7]
```diff
@@ -12,6 +12,8 @@ def _output_dir_4saving(output_dir=None, rootfile=None):
             # if rootfile is specified, use it's directory
             output_dir = os.path.dirname(rootfile)
 
+    if output_dir=='' : output_dir = './'
```
Comment (juhuntenburg) [inline nighres/utils.py]: why is this necessary? isn't this what line 10 does (but in an cross-platform way)? Why would output_dir ever be set to an empty string (instead of to None as is default in all functions)?

File: nighres/utils.py [hunk 8]
```diff
@@ -40,9 +42,9 @@ def _fname_4saving(file_name=None, rootfile=None, suffix=None):
         # use its file_name
         if isinstance(rootfile, basestring):
             file_name = os.path.basename(rootfile)
-            print(("You have not specified a file_name. We will use the "
-                   "name of your input ({0}) as a base name for saving "
-                   "outputs.").format(file_name))
+            #print(("You have not specified a file_name. We will use the "
+            #       "name of your input ({0}) as a base name for saving "
+            #       "outputs.").format(file_name))
```
Comment (juhuntenburg) [inline nighres/utils.py]: Why removing this notice?

File: nighres/registration/embedded_syn.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+T=3
+
+def embedded_syn(source_image, target_image, coarse_iterations=40, medium_iterations=50, fine_iterations=40,
+					run_affine_first=False, cost_function='MutualInformation', interpolation='NearestNeighbor',
+                    save_data=False, output_dir=None,
+                    file_name=None):
+    """ Embedded SyN
+
+    Runs the Symmetric Normalization (SyN) algorithm of ANTs and formats the output deformations
+    into voxel coordinate mappings as used in CBSTools registration and transformation routines.
```
Comment (juhuntenburg) [inline nighres/registration/embedded_syn.py]: We should add that this requires installation of Nipype and ANTs

**Comment by juhuntenburg** [comment]
build-nodependencies.sh works fine for me! I think we can safely rename it to build.sh (to replace the old build.sh and avoid confusion). As the other parts of the PR might still need a little work and I would like to merge this quickly, could you do the following:
1. pull the latest changes into your local master (pull upstream master)
2. checkout a completely new branch
3. replace the build.sh with the version here (just copy it somewhere else before leaving your devel branch)
4. Make new PR that only contains that change (and remove the build-nodependencies.sh from this PR)?

I could also do it quickly, but its not so nice because it doesn't track that this is your contribution. Should be really quick to do and I can merge it immediately

**Comment by piloubazin** [comment]
Done! You should get a new PR by now. I'll look over these questions later on, but we could split again the branch to address the more pressing points (i.e. CRUISE) quickly

File: nighres/utils.py (hunk 7 above)
Comment (piloubazin) [reply nighres/utils.py]: it didn't work, as you need some directory to be set in order to add to it for creating filenames (I think the empty string and None have the same problem, may be in fact the same thing)

File: nighres/utils.py (hunk 8 above)
Comment (piloubazin) [reply nighres/utils.py]: once you input and output more that a couple of files this becomes very irritating, to be honest.

File: nighres/registration/embedded_syn.py (hunk 9 above)
Comment (piloubazin) [reply nighres/registration/embedded_syn.py]: I could remove the Nipype layer if needed (in the end it doesn't bring much in this case)

**Comment by piloubazin** [comment]
So I'm fine splitting things to make it more modular (still getting used to this branching business), but how do I remove pieces from the PR? or do I just discard it in the end and make several new branches,each with their PR?

File: nighres/registration/embedded_syn.py (hunk 9 above)
Comment (juhuntenburg) [reply nighres/registration/embedded_syn.py]: Ok sounds good, less dependencies

File: nighres/utils.py (hunk 7 above)
Comment (juhuntenburg) [reply nighres/utils.py]: Still not sure what you mean. If no output_dir, nor a rootfile is given, the expected behaviour is to use the current working directory as output_dir (see line 10).  What problem did you run into exactly?

File: nighres/utils.py (hunk 8 above)
Comment (juhuntenburg) [reply nighres/utils.py]: ok, makes sense

**Comment by juhuntenburg** [comment]
@piloubazin both is possible. You can move each of your features to a new branch. You can also keep this PR and remove things. A PR automatically updates when you change your remote branch that you made the PR from. So all you need to do is make your changes locally and push to your origin devel branch. There are two cases:
1. If you want to remove a new file you added -- just delete it in your local repo on the devel branch, commit the removing of the file, and then push to your remote branch
2. If you want to reset a file that you changed to the version it has on the master branch you can do 
`git checkout origin/master filename`, then you still need to commit the change and again push to your remote branch

Does this help?

File: nighres/filter/bandpass_filtering.py [hunk 10]
```diff
@@ -59,7 +59,7 @@ def bandpass_filtering(time_series, low_frequency=0.01, high_frequency=0.1,
 
         filtered_file = _fname_4saving(file_name=file_name,
                                    rootfile=time_series,
-                                   suffix='bpf')
+                                   suffix='bpf_img')
 
      # get dimensions and resolution from second inversion image
     img = load_volume(time_series)
```
[the remark this answers is not in the document]
Comment (piloubazin) [reply nighres/filter/bandpass_filtering.py]: can you just push your version, then? I'll pull from there and make the needed fixes (low priority)

File: nighres/utils.py (hunk 7 above)
Comment (piloubazin) [reply nighres/utils.py]: my bad method to create file paths with dir+name failed, but I think I have the solution now using os.path

File: nighres/registration/embedded_syn.py (hunk 5 above)
Comment (piloubazin) [reply nighres/registration/embedded_syn.py]: actually, output_dir should always be set to something meaningful (if you don't give any, the default is to use the current folder, right? so io._output_dir_4saving should return that in case of None input. No?

**Comment by juhuntenburg** [comment]
Closing because cruise was merged with other PR
